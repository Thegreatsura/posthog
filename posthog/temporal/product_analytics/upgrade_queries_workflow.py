import dataclasses
import datetime as dt
import json
import textwrap

from django.db import connection
import temporalio.workflow
from posthog.models import Insight
from posthog.schema_migrations.upgrade import upgrade
from posthog.temporal.common.base import PostHogWorkflow
from posthog.temporal.common.logger import get_internal_logger
from posthog.schema_migrations import LATEST_VERSIONS


def _clause(kind: str, version: int) -> str:
    template = """
        query @? '$.** ? (
            @.kind == "{kind}" &&
            (!exists(@.v) || @.v == null || @.v <= {version})
        )'"""
    return textwrap.dedent(template.format(kind=kind, version=version)).strip()


@dataclasses.dataclass(frozen=True)
class GetInsightsToMigrateActivityInputs:
    """Inputs for the get insights to migrate activity."""

    batch_size: int = dataclasses.field(default=100)


@temporalio.activity.defn
def get_insights_to_migrate(inputs: GetInsightsToMigrateActivityInputs) -> list[int]:
    clauses = [_clause(k, v) for k, v in sorted(LATEST_VERSIONS.items())]
    where_body = ("\n   OR  ").join(clauses)
    sql = f"""
        SELECT DISTINCT id
        FROM posthog_dashboarditem
        WHERE {where_body}
        LIMIT {inputs.batch_size};
    """

    with connection.cursor() as cur:
        cur.execute(sql)
        ids = [row[0] for row in cur.fetchall()]

    return ids


@dataclasses.dataclass(frozen=True)
class MigrateInsightsBatchActivityInputs:
    """Inputs for the migrate insights batch activity."""

    insight_ids: list[int] = dataclasses.field()


@temporalio.activity.defn
def migrate_insights_batch(inputs: MigrateInsightsBatchActivityInputs) -> None:
    """Migrate a batch of insights to the latest version."""
    logger = get_internal_logger()

    insights = Insight.objects.filter(id__in=inputs.insight_ids)

    for insight in insights:
        try:
            insight.query = upgrade(insight.query)
            insight.save()
        except Exception as e:
            logger.exception(f"Error migrating insight {insight.id}: {str(e)}")


@dataclasses.dataclass(frozen=True)
class UpgradeQueriesWorkflowInputs:
    """Inputs for the upgrade queries workflow."""

    batch_size: int = dataclasses.field(default=100)


@temporalio.workflow.defn(name="upgrade-queries")
class UpgradeQueriesWorkflow(PostHogWorkflow):
    @staticmethod
    def parse_inputs(inputs: list[str]) -> UpgradeQueriesWorkflowInputs:
        """Parse inputs from the management command CLI."""
        loaded = json.loads(inputs[0])
        return UpgradeQueriesWorkflowInputs(**loaded)

    @temporalio.workflow.run
    async def run(self, inputs: UpgradeQueriesWorkflowInputs) -> None:
        insight_ids = await temporalio.workflow.execute_activity(
            get_insights_to_migrate,
            GetInsightsToMigrateActivityInputs(batch_size=inputs.batch_size),
            start_to_close_timeout=dt.timedelta(minutes=5),
            retry_policy=temporalio.common.RetryPolicy(
                initial_interval=dt.timedelta(minutes=5),
                maximum_interval=dt.timedelta(minutes=60),
                maximum_attempts=3,
            ),
        )

        while len(insight_ids) > 0:
            await temporalio.workflow.execute_activity(
                migrate_insights_batch,
                MigrateInsightsBatchActivityInputs(insight_ids=insight_ids),
                start_to_close_timeout=dt.timedelta(minutes=10),
                retry_policy=temporalio.common.RetryPolicy(
                    initial_interval=dt.timedelta(minutes=10),
                    maximum_interval=dt.timedelta(minutes=60),
                    maximum_attempts=3,
                ),
            )
            insight_ids = await temporalio.workflow.execute_activity(
                get_insights_to_migrate,
                GetInsightsToMigrateActivityInputs(batch_size=inputs.batch_size),
                start_to_close_timeout=dt.timedelta(minutes=5),
                retry_policy=temporalio.common.RetryPolicy(
                    initial_interval=dt.timedelta(minutes=5),
                    maximum_interval=dt.timedelta(minutes=60),
                    maximum_attempts=3,
                ),
            )
