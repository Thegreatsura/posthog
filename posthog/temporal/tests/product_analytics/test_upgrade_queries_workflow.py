import pytest

from posthog.models.insight import Insight
from posthog.schema import NodeKind
from posthog.schema_migrations import LATEST_VERSIONS, MIGRATIONS, SchemaMigration
from posthog.temporal.product_analytics.upgrade_queries_workflow import (
    GetInsightsToMigrateActivityInputs,
    MigrateInsightsBatchActivityInputs,
    get_insights_to_migrate,
    migrate_insights_batch,
)
from posthog.test.base import QueryMatchingTest, snapshot_postgres_queries_context


class TrendsMigration(SchemaMigration):
    targets = {NodeKind.TRENDS_QUERY: 5}

    def transform(self, query):
        query["interval"] = "day"
        return query


@pytest.fixture(autouse=True)
def setup_migrations():
    LATEST_VERSIONS.clear()
    MIGRATIONS.clear()

    MIGRATIONS[NodeKind.TRENDS_QUERY] = {1: TrendsMigration()}
    # MIGRATIONS[NodeKind.EVENTS_NODE] = {1: EventsNodeMigration()}
    LATEST_VERSIONS[NodeKind.INSIGHT_VIZ_NODE] = 3
    LATEST_VERSIONS[NodeKind.TRENDS_QUERY] = 5
    LATEST_VERSIONS[NodeKind.EVENTS_NODE] = 7

    yield


def setup_insights(team):
    # all versions satisfied
    i1 = Insight.objects.create(
        query={
            "kind": "InsightVizNode",
            "source": {
                "kind": "TrendsQuery",
                "series": [{"kind": "EventsNode", "event": "$pageview", "v": 8}],
                "v": 6,
            },
            "v": 4,
        },
        team=team,
    )

    # matching top-level node (InsightVizNode)
    i2 = Insight.objects.create(
        query={
            "kind": "InsightVizNode",
            "source": {
                "kind": "TrendsQuery",
                "series": [{"kind": "EventsNode", "event": "$pageview", "v": 8}],
                "v": 6,
            },
            "v": 3,
        },
        team=team,
    )

    # matching node in nested dict (TrendsQuery)
    i3 = Insight.objects.create(
        query={
            "kind": "InsightVizNode",
            "source": {
                "kind": "TrendsQuery",
                "series": [{"kind": "EventsNode", "event": "$pageview", "v": 8}],
                "v": 5,
            },
            "v": 4,
        },
        team=team,
    )

    # matching node in nested list (EventsNode)
    i4 = Insight.objects.create(
        query={
            "kind": "InsightVizNode",
            "source": {
                "kind": "TrendsQuery",
                "series": [{"kind": "EventsNode", "event": "$pageview", "v": 7}],
                "v": 6,
            },
            "v": 4,
        },
        team=team,
    )

    # no query
    i5 = Insight.objects.create(team=team)

    # kind without migration (DataTableNode)
    i6 = Insight.objects.create(
        query={
            "kind": "DataTableNode",
        },
        team=team,
    )

    # no version (InsightVizNode)
    i7 = Insight.objects.create(
        query={
            "kind": "InsightVizNode",
            "source": {
                "kind": "TrendsQuery",
                "series": [{"kind": "EventsNode", "event": "$pageview", "v": 8}],
                "v": 6,
            },
        },
        team=team,
    )

    # none version (InsightVizNode)
    i8 = Insight.objects.create(
        query={
            "kind": "InsightVizNode",
            "source": {
                "kind": "TrendsQuery",
                "series": [{"kind": "EventsNode", "event": "$pageview", "v": 8}],
                "v": 6,
            },
            "v": None,
        },
        team=team,
    )

    return i1, i2, i3, i4, i5, i6, i7, i8


class TestUpgradeQueriesWorkflow(QueryMatchingTest):
    @pytest.mark.django_db(transaction=True)
    def test_get_insights_to_migrate_activity(self, activity_environment, team):
        i1, i2, i3, i4, i5, i6, i7, i8 = setup_insights(team)
        inputs = GetInsightsToMigrateActivityInputs()

        with snapshot_postgres_queries_context(self):
            result = activity_environment.run(get_insights_to_migrate, inputs)

        expected_ids = [i2.id, i3.id, i4.id, i7.id, i8.id]
        assert sorted(result), expected_ids

    @pytest.mark.django_db(transaction=True)
    def test_migrate_insights_batch_activity(self, activity_environment, team):
        i1, i2, i3, i4, i5, i6, i7, i8 = setup_insights(team)
        inputs = MigrateInsightsBatchActivityInputs(
            insight_ids=[i2.id, i3.id, i4.id, i7.id, i8.id],
        )

        result = activity_environment.run(migrate_insights_batch, inputs)

        assert i3.query == "s"
