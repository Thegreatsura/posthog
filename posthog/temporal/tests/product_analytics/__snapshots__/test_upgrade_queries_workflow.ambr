# serializer version: 1
# name: TestUpgradeQueriesWorkflow.test_get_insights_to_migrate_activity
  '''
  WITH latest(kind, v_latest) AS (
                                  VALUES ('InsightVizNode',
                                          4),('TrendsQuery',
                                              6),('EventsNode',
                                                  8))
  SELECT DISTINCT i.id
  FROM posthog_dashboarditem AS i
  CROSS JOIN latest AS l
  WHERE jsonb_path_exists(i.query, format($$
                  $.** ? (
                      @.kind == "%s" &&    /* kind matches */
                      (
                          !exists(@.v)     /* no v property */
                          || @.v == null   /* or v:null */
                          || @.v <= %s     /* or version â‰¤ latest */
                      )
                  )
                  $$, l.kind, l.v_latest)::jsonpath)
  LIMIT 100;
  '''
# ---
